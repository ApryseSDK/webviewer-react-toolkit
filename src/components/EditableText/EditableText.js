'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.EditableText = void 0;
var tslib_1 = require('tslib');
var classnames_1 = tslib_1.__importDefault(require('classnames'));
var react_1 = tslib_1.__importStar(require('react'));
var hooks_1 = require('../../hooks');
var ClickableDiv_1 = require('../ClickableDiv');
exports.EditableText = (0, react_1.forwardRef)(function(_a, ref) {
  var controlledValue = _a.value,
    controlledEditMode = _a.editMode,
    onEdit = _a.onEdit,
    onSave = _a.onSave,
    onCancel = _a.onCancel,
    className = _a.className,
    disabled = _a.disabled,
    locked = _a.locked,
    onRenderText = _a.onRenderText,
    placeholder = _a.placeholder,
    centerText = _a.centerText,
    bordered = _a.bordered,
    clickableDivProps = tslib_1.__rest(_a, [
      'value',
      'editMode',
      'onEdit',
      'onSave',
      'onCancel',
      'className',
      'disabled',
      'locked',
      'onRenderText',
      'placeholder',
      'centerText',
      'bordered',
    ]);
  var inputRef = (0, react_1.useRef)(null);
  var buttonRef = (0, react_1.useRef)(null);
  (0, react_1.useImperativeHandle)(ref, function() {
    return buttonRef.current;
  });
  // Use state if controlled edit mode not provided.
  var _b = (0, react_1.useState)(false),
    stateEditMode = _b[0],
    setStateEditMode = _b[1];
  var editMode =
    controlledEditMode !== null && controlledEditMode !== void 0
      ? controlledEditMode
      : stateEditMode;
  // Use state if controlled value not provided.
  var _c = (0, react_1.useState)(''),
    stateValue = _c[0],
    setStateValue = _c[1];
  var value =
    controlledValue !== null && controlledValue !== void 0
      ? controlledValue
      : stateValue;
  // Keep edit value in sync with value (controlled or state) and reset when
  // edit mode is cancelled.
  var _d = (0, react_1.useState)(value),
    editValue = _d[0],
    setEditValue = _d[1];
  (0, react_1.useEffect)(
    function() {
      return setEditValue(value);
    },
    [editMode, value],
  );
  var _e = (0, react_1.useState)(false),
    noFocusTransition = _e[0],
    setNoFocusTransition = _e[1];
  // Focus input whenever edit mode is enabled, and button when disabled.
  var firstRender = (0, react_1.useRef)(true);
  (0, react_1.useEffect)(
    function() {
      var _a, _b;
      if (firstRender.current) {
        firstRender.current = false;
      } else if (editMode) {
        (_a = inputRef.current) === null || _a === void 0 ? void 0 : _a.focus();
        setNoFocusTransition(true);
      } else {
        (_b = buttonRef.current) === null || _b === void 0
          ? void 0
          : _b.focus();
        setNoFocusTransition(false);
      }
    },
    [editMode],
  );
  var handleOnEdit = function() {
    if (controlledEditMode === undefined) setStateEditMode(true);
    onEdit === null || onEdit === void 0 ? void 0 : onEdit();
  };
  var handleOnCancel = function() {
    if (controlledEditMode === undefined) setStateEditMode(false);
    setEditValue(value);
    onCancel === null || onCancel === void 0 ? void 0 : onCancel(value);
  };
  var handleOnSave = function() {
    if (controlledValue === undefined) setStateValue(editValue);
    if (controlledEditMode === undefined) setStateEditMode(false);
    onSave === null || onSave === void 0 ? void 0 : onSave(editValue);
  };
  var handleOnKeyPress = function(event) {
    if (event.key === 'Enter') {
      handleOnSave();
      event.stopPropagation();
    }
  };
  var handleOnKeyDown = function(event) {
    if (event.key === 'Escape') {
      handleOnCancel();
    }
    event.stopPropagation();
  };
  var _f = (0, react_1.useMemo)(
      function() {
        var renderedValue = onRenderText ? onRenderText(value) : value;
        if (placeholder && !renderedValue) return [placeholder, true];
        return [renderedValue, false];
      },
      [onRenderText, placeholder, value],
    ),
    valueToDisplay = _f[0],
    isPlaceholder = _f[1];
  var isUserTabbing = (0, hooks_1.useAccessibleFocus)();
  var editableTextClass = (0, classnames_1.default)(
    'ui__base ui__editableText',
    {
      'ui__editableText--disabled': disabled,
      'ui__editableText--locked': locked,
      'ui__editableText--centerText': centerText,
      'ui__editableText--bordered': bordered,
    },
    className,
  );
  var buttonClass = (0, classnames_1.default)('ui__editableText__button', {
    'ui__editableText__button--placeholder': isPlaceholder,
    'ui__editableText__button--noFocusTransition': noFocusTransition,
  });
  var fieldClass = (0, classnames_1.default)('ui__editableText__field', {
    'ui__editableText__field--tabbing': isUserTabbing,
  });
  return react_1.default.createElement(
    'div',
    { className: editableTextClass },
    editMode
      ? react_1.default.createElement('input', {
          className: fieldClass,
          value: editValue,
          onChange: function(e) {
            return setEditValue(e.target.value);
          },
          onKeyPress: handleOnKeyPress,
          onKeyDown: handleOnKeyDown,
          ref: inputRef,
          onBlur: handleOnSave,
          onClick: function(e) {
            return e.stopPropagation();
          },
        })
      : react_1.default.createElement(
          ClickableDiv_1.ClickableDiv,
          tslib_1.__assign({}, clickableDivProps, {
            className: buttonClass,
            disabled: disabled || locked,
            onClick: handleOnEdit,
            ref: buttonRef,
            usePointer: true,
          }),
          react_1.default.createElement('span', null, valueToDisplay),
        ),
  );
});
//# sourceMappingURL=EditableText.js.map
