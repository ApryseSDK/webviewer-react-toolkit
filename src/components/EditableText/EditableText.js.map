{"version":3,"file":"EditableText.js","sourceRoot":"","sources":["EditableText.tsx"],"names":[],"mappings":";;;;AAAA,kEAAoC;AACpC,qDAAoH;AACpH,qCAAiD;AACjD,gDAAkE;AA0DrD,QAAA,YAAY,GAAG,IAAA,kBAAU,EACpC,UACE,EAcC,EACD,GAAG;IAdD,IAAO,eAAe,WAAA,EACZ,kBAAkB,cAAA,EAC5B,MAAM,YAAA,EACN,MAAM,YAAA,EACN,QAAQ,cAAA,EACR,SAAS,eAAA,EACT,QAAQ,cAAA,EACR,MAAM,YAAA,EACN,YAAY,kBAAA,EACZ,WAAW,iBAAA,EACX,UAAU,gBAAA,EACV,QAAQ,cAAA,EACL,iBAAiB,sBAbtB,iJAcC,CADqB;IAItB,IAAM,QAAQ,GAAG,IAAA,cAAM,EAAmB,IAAI,CAAC,CAAC;IAEhD,IAAM,SAAS,GAAG,IAAA,cAAM,EAAiB,IAAI,CAAC,CAAC;IAC/C,IAAA,2BAAmB,EAAC,GAAG,EAAE,cAAM,OAAA,SAAS,CAAC,OAAyB,EAAnC,CAAmC,CAAC,CAAC;IAEpE,kDAAkD;IAC5C,IAAA,KAAoC,IAAA,gBAAQ,EAAC,KAAK,CAAC,EAAlD,aAAa,QAAA,EAAE,gBAAgB,QAAmB,CAAC;IAC1D,IAAM,QAAQ,GAAG,kBAAkB,aAAlB,kBAAkB,cAAlB,kBAAkB,GAAI,aAAa,CAAC;IAErD,8CAA8C;IACxC,IAAA,KAA8B,IAAA,gBAAQ,EAAC,EAAE,CAAC,EAAzC,UAAU,QAAA,EAAE,aAAa,QAAgB,CAAC;IACjD,IAAM,KAAK,GAAG,eAAe,aAAf,eAAe,cAAf,eAAe,GAAI,UAAU,CAAC;IAE5C,0EAA0E;IAC1E,0BAA0B;IACpB,IAAA,KAA4B,IAAA,gBAAQ,EAAC,KAAK,CAAC,EAA1C,SAAS,QAAA,EAAE,YAAY,QAAmB,CAAC;IAClD,IAAA,iBAAS,EAAC,cAAM,OAAA,YAAY,CAAC,KAAK,CAAC,EAAnB,CAAmB,EAAE,CAAC,QAAQ,EAAE,KAAK,CAAC,CAAC,CAAC;IAElD,IAAA,KAA4C,IAAA,gBAAQ,EAAC,KAAK,CAAC,EAA1D,iBAAiB,QAAA,EAAE,oBAAoB,QAAmB,CAAC;IAElE,uEAAuE;IACvE,IAAM,WAAW,GAAG,IAAA,cAAM,EAAC,IAAI,CAAC,CAAC;IACjC,IAAA,iBAAS,EAAC;;QACR,IAAI,WAAW,CAAC,OAAO,EAAE;YACvB,WAAW,CAAC,OAAO,GAAG,KAAK,CAAC;SAC7B;aAAM,IAAI,QAAQ,EAAE;YACnB,MAAA,QAAQ,CAAC,OAAO,0CAAE,KAAK,EAAE,CAAC;YAC1B,oBAAoB,CAAC,IAAI,CAAC,CAAC;SAC5B;aAAM;YACL,MAAA,SAAS,CAAC,OAAO,0CAAE,KAAK,EAAE,CAAC;YAC3B,oBAAoB,CAAC,KAAK,CAAC,CAAC;SAC7B;IACH,CAAC,EAAE,CAAC,QAAQ,CAAC,CAAC,CAAC;IAEf,IAAM,YAAY,GAAG;QACnB,IAAI,kBAAkB,KAAK,SAAS;YAAE,gBAAgB,CAAC,IAAI,CAAC,CAAC;QAC7D,MAAM,aAAN,MAAM,uBAAN,MAAM,EAAI,CAAC;IACb,CAAC,CAAC;IAEF,IAAM,cAAc,GAAG;QACrB,IAAI,kBAAkB,KAAK,SAAS;YAAE,gBAAgB,CAAC,KAAK,CAAC,CAAC;QAC9D,YAAY,CAAC,KAAK,CAAC,CAAC;QACpB,QAAQ,aAAR,QAAQ,uBAAR,QAAQ,CAAG,KAAK,CAAC,CAAC;IACpB,CAAC,CAAC;IAEF,IAAM,YAAY,GAAG;QACnB,IAAI,eAAe,KAAK,SAAS;YAAE,aAAa,CAAC,SAAS,CAAC,CAAC;QAC5D,IAAI,kBAAkB,KAAK,SAAS;YAAE,gBAAgB,CAAC,KAAK,CAAC,CAAC;QAC9D,MAAM,aAAN,MAAM,uBAAN,MAAM,CAAG,SAAS,CAAC,CAAC;IACtB,CAAC,CAAC;IAEF,IAAM,gBAAgB,GAAG,UAAC,KAAsC;QAC9D,IAAI,KAAK,CAAC,GAAG,KAAK,OAAO,EAAE;YACzB,YAAY,EAAE,CAAC;YACf,KAAK,CAAC,eAAe,EAAE,CAAC;SACzB;IACH,CAAC,CAAC;IAEF,IAAM,eAAe,GAAG,UAAC,KAAsC;QAC7D,IAAI,KAAK,CAAC,GAAG,KAAK,QAAQ,EAAE;YAC1B,cAAc,EAAE,CAAC;SAClB;QACD,KAAK,CAAC,eAAe,EAAE,CAAC;IAC1B,CAAC,CAAC;IAEI,IAAA,KAAkC,IAAA,eAAO,EAAC;QAC9C,IAAM,aAAa,GAAG,YAAY,CAAC,CAAC,CAAC,YAAY,CAAC,KAAK,CAAC,CAAC,CAAC,CAAC,KAAK,CAAC;QACjE,IAAI,WAAW,IAAI,CAAC,aAAa;YAAE,OAAO,CAAC,WAAW,EAAE,IAAI,CAAC,CAAC;QAC9D,OAAO,CAAC,aAAa,EAAE,KAAK,CAAC,CAAC;IAChC,CAAC,EAAE,CAAC,YAAY,EAAE,WAAW,EAAE,KAAK,CAAC,CAAC,EAJ/B,cAAc,QAAA,EAAE,aAAa,QAIE,CAAC;IAEvC,IAAM,aAAa,GAAG,IAAA,0BAAkB,GAAE,CAAC;IAE3C,IAAM,iBAAiB,GAAG,IAAA,oBAAU,EAClC,2BAA2B,EAC3B;QACE,4BAA4B,EAAE,QAAQ;QACtC,0BAA0B,EAAE,MAAM;QAClC,8BAA8B,EAAE,UAAU;QAC1C,4BAA4B,EAAE,QAAQ;KACvC,EACD,SAAS,CACV,CAAC;IAEF,IAAM,WAAW,GAAG,IAAA,oBAAU,EAAC,0BAA0B,EAAE;QACzD,uCAAuC,EAAE,aAAa;QACtD,6CAA6C,EAAE,iBAAiB;KACjE,CAAC,CAAC;IAEH,IAAM,UAAU,GAAG,IAAA,oBAAU,EAAC,yBAAyB,EAAE;QACvD,kCAAkC,EAAE,aAAa;KAClD,CAAC,CAAC;IAEH,OAAO,CACL,uCAAK,SAAS,EAAE,iBAAiB,IAC9B,QAAQ,CAAC,CAAC,CAAC,CACV,yCACE,SAAS,EAAE,UAAU,EACrB,KAAK,EAAE,SAAS,EAChB,QAAQ,EAAE,UAAC,CAAC,IAAK,OAAA,YAAY,CAAC,CAAC,CAAC,MAAM,CAAC,KAAK,CAAC,EAA5B,CAA4B,EAC7C,UAAU,EAAE,gBAAgB,EAC5B,SAAS,EAAE,eAAe,EAC1B,GAAG,EAAE,QAAQ,EACb,MAAM,EAAE,YAAY,EACpB,OAAO,EAAE,UAAC,CAAC,IAAK,OAAA,CAAC,CAAC,eAAe,EAAE,EAAnB,CAAmB,GACnC,CACH,CAAC,CAAC,CAAC,CACF,8BAAC,2BAAY,uBACP,iBAAiB,IACrB,SAAS,EAAE,WAAW,EACtB,QAAQ,EAAE,QAAQ,IAAI,MAAM,EAC5B,OAAO,EAAE,YAAY,EACrB,GAAG,EAAE,SAAS,EACd,UAAU;QAEV,4CAAO,cAAc,CAAQ,CAChB,CAChB,CACG,CACP,CAAC;AACJ,CAAC,CACF,CAAC"}