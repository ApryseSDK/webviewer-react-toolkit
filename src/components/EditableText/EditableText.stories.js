'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.FullyControlled = exports.ControlledEditMode = exports.ControlledValue = exports.WithOnRenderText = exports.Basic = void 0;
var tslib_1 = require('tslib');
var addon_knobs_1 = require('@storybook/addon-knobs');
var react_1 = tslib_1.__importDefault(require('react'));
var action_1 = require('../../storybook-helpers/action/action');
var EditableText_1 = require('../EditableText');
var README_md_1 = tslib_1.__importDefault(require('./README.md'));
exports.default = {
  title: 'Components/EditableText',
  component: EditableText_1.EditableText,
  parameters: { readme: README_md_1.default },
};
var Basic = function() {
  return react_1.default.createElement(EditableText_1.EditableText, {
    centerText: (0, addon_knobs_1.boolean)('centerText', false),
    bordered: (0, addon_knobs_1.boolean)('bordered', true),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    locked: (0, addon_knobs_1.boolean)('locked', false),
    placeholder: (0, addon_knobs_1.text)('placeholder', ''),
    onSave: (0, action_1.action)('onSave'),
    onCancel: (0, action_1.action)('onCancel'),
    onEdit: (0, action_1.action)('onEdit'),
  });
};
exports.Basic = Basic;
var WithOnRenderText = function() {
  return react_1.default.createElement(EditableText_1.EditableText, {
    onRenderText: function(value) {
      return value ? '[[ '.concat(value.toUpperCase(), ' ]]') : '';
    },
    centerText: (0, addon_knobs_1.boolean)('centerText', false),
    bordered: (0, addon_knobs_1.boolean)('bordered', true),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    locked: (0, addon_knobs_1.boolean)('locked', false),
    placeholder: (0, addon_knobs_1.text)('placeholder', ''),
    onSave: (0, action_1.action)('onSave'),
    onCancel: (0, action_1.action)('onCancel'),
    onEdit: (0, action_1.action)('onEdit'),
  });
};
exports.WithOnRenderText = WithOnRenderText;
var ControlledValue = function() {
  return react_1.default.createElement(EditableText_1.EditableText, {
    value: (0, addon_knobs_1.text)('value', ''),
    centerText: (0, addon_knobs_1.boolean)('centerText', false),
    bordered: (0, addon_knobs_1.boolean)('bordered', true),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    locked: (0, addon_knobs_1.boolean)('locked', false),
    placeholder: (0, addon_knobs_1.text)('placeholder', ''),
    onSave: (0, action_1.action)('onSave'),
    onCancel: (0, action_1.action)('onCancel'),
    onEdit: (0, action_1.action)('onEdit'),
  });
};
exports.ControlledValue = ControlledValue;
var ControlledEditMode = function() {
  return react_1.default.createElement(EditableText_1.EditableText, {
    editMode: (0, addon_knobs_1.boolean)('editMode', false),
    centerText: (0, addon_knobs_1.boolean)('centerText', false),
    bordered: (0, addon_knobs_1.boolean)('bordered', true),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    locked: (0, addon_knobs_1.boolean)('locked', false),
    placeholder: (0, addon_knobs_1.text)('placeholder', ''),
    onSave: (0, action_1.action)('onSave'),
    onCancel: (0, action_1.action)('onCancel'),
    onEdit: (0, action_1.action)('onEdit'),
  });
};
exports.ControlledEditMode = ControlledEditMode;
var FullyControlled = function() {
  return react_1.default.createElement(EditableText_1.EditableText, {
    value: (0, addon_knobs_1.text)('value', ''),
    editMode: (0, addon_knobs_1.boolean)('editMode', false),
    centerText: (0, addon_knobs_1.boolean)('centerText', false),
    bordered: (0, addon_knobs_1.boolean)('bordered', true),
    disabled: (0, addon_knobs_1.boolean)('disabled', false),
    locked: (0, addon_knobs_1.boolean)('locked', false),
    placeholder: (0, addon_knobs_1.text)('placeholder', ''),
    onSave: (0, action_1.action)('onSave'),
    onCancel: (0, action_1.action)('onCancel'),
    onEdit: (0, action_1.action)('onEdit'),
  });
};
exports.FullyControlled = FullyControlled;
//# sourceMappingURL=EditableText.stories.js.map
