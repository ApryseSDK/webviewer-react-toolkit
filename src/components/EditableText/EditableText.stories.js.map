{"version":3,"file":"EditableText.stories.js","sourceRoot":"","sources":["EditableText.stories.tsx"],"names":[],"mappings":";;;;AAAA,sDAAuD;AACvD,wDAA0B;AAC1B,gEAA+D;AAC/D,gDAA+C;AAC/C,kEAAiC;AAEjC,kBAAe,EAAE,KAAK,EAAE,yBAAyB,EAAE,SAAS,EAAE,2BAAY,EAAE,UAAU,EAAE,EAAE,MAAM,qBAAA,EAAE,EAAE,CAAC;AAE9F,IAAM,KAAK,GAAG,cAAM,OAAA,CACzB,8BAAC,2BAAY,IACX,UAAU,EAAE,IAAA,qBAAO,EAAC,YAAY,EAAE,KAAK,CAAC,EACxC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,IAAI,CAAC,EACnC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,MAAM,EAAE,IAAA,qBAAO,EAAC,QAAQ,EAAE,KAAK,CAAC,EAChC,WAAW,EAAE,IAAA,kBAAI,EAAC,aAAa,EAAE,EAAE,CAAC,EACpC,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,EACxB,QAAQ,EAAE,IAAA,eAAM,EAAC,UAAU,CAAC,EAC5B,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,GACxB,CACH,EAX0B,CAW1B,CAAC;AAXW,QAAA,KAAK,SAWhB;AAEK,IAAM,gBAAgB,GAAG,cAAM,OAAA,CACpC,8BAAC,2BAAY,IACX,YAAY,EAAE,UAAC,KAAK,IAAK,OAAA,CAAC,KAAK,CAAC,CAAC,CAAC,aAAM,KAAK,CAAC,WAAW,EAAE,QAAK,CAAC,CAAC,CAAC,EAAE,CAAC,EAA7C,CAA6C,EACtE,UAAU,EAAE,IAAA,qBAAO,EAAC,YAAY,EAAE,KAAK,CAAC,EACxC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,IAAI,CAAC,EACnC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,MAAM,EAAE,IAAA,qBAAO,EAAC,QAAQ,EAAE,KAAK,CAAC,EAChC,WAAW,EAAE,IAAA,kBAAI,EAAC,aAAa,EAAE,EAAE,CAAC,EACpC,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,EACxB,QAAQ,EAAE,IAAA,eAAM,EAAC,UAAU,CAAC,EAC5B,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,GACxB,CACH,EAZqC,CAYrC,CAAC;AAZW,QAAA,gBAAgB,oBAY3B;AAEK,IAAM,eAAe,GAAG,cAAM,OAAA,CACnC,8BAAC,2BAAY,IACX,KAAK,EAAE,IAAA,kBAAI,EAAC,OAAO,EAAE,EAAE,CAAC,EACxB,UAAU,EAAE,IAAA,qBAAO,EAAC,YAAY,EAAE,KAAK,CAAC,EACxC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,IAAI,CAAC,EACnC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,MAAM,EAAE,IAAA,qBAAO,EAAC,QAAQ,EAAE,KAAK,CAAC,EAChC,WAAW,EAAE,IAAA,kBAAI,EAAC,aAAa,EAAE,EAAE,CAAC,EACpC,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,EACxB,QAAQ,EAAE,IAAA,eAAM,EAAC,UAAU,CAAC,EAC5B,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,GACxB,CACH,EAZoC,CAYpC,CAAC;AAZW,QAAA,eAAe,mBAY1B;AAEK,IAAM,kBAAkB,GAAG,cAAM,OAAA,CACtC,8BAAC,2BAAY,IACX,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,UAAU,EAAE,IAAA,qBAAO,EAAC,YAAY,EAAE,KAAK,CAAC,EACxC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,IAAI,CAAC,EACnC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,MAAM,EAAE,IAAA,qBAAO,EAAC,QAAQ,EAAE,KAAK,CAAC,EAChC,WAAW,EAAE,IAAA,kBAAI,EAAC,aAAa,EAAE,EAAE,CAAC,EACpC,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,EACxB,QAAQ,EAAE,IAAA,eAAM,EAAC,UAAU,CAAC,EAC5B,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,GACxB,CACH,EAZuC,CAYvC,CAAC;AAZW,QAAA,kBAAkB,sBAY7B;AAEK,IAAM,eAAe,GAAG,cAAM,OAAA,CACnC,8BAAC,2BAAY,IACX,KAAK,EAAE,IAAA,kBAAI,EAAC,OAAO,EAAE,EAAE,CAAC,EACxB,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,UAAU,EAAE,IAAA,qBAAO,EAAC,YAAY,EAAE,KAAK,CAAC,EACxC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,IAAI,CAAC,EACnC,QAAQ,EAAE,IAAA,qBAAO,EAAC,UAAU,EAAE,KAAK,CAAC,EACpC,MAAM,EAAE,IAAA,qBAAO,EAAC,QAAQ,EAAE,KAAK,CAAC,EAChC,WAAW,EAAE,IAAA,kBAAI,EAAC,aAAa,EAAE,EAAE,CAAC,EACpC,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,EACxB,QAAQ,EAAE,IAAA,eAAM,EAAC,UAAU,CAAC,EAC5B,MAAM,EAAE,IAAA,eAAM,EAAC,QAAQ,CAAC,GACxB,CACH,EAboC,CAapC,CAAC;AAbW,QAAA,eAAe,mBAa1B"}