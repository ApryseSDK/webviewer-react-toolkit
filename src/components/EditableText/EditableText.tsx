import classnames from 'classnames';
import React, { forwardRef, KeyboardEvent, useCallback, useEffect, useRef, useState, useMemo } from 'react';
import ClickableDiv from '../ClickableDiv';

export interface EditableTextProps {
  /**
   * If provided, value becomes controlled. The component will request to change
   * the value with the onSave callback, and this must be set to change.
   */
  value?: string;
  /**
   * If provided, edit mode becomes controlled. The component will request to
   * enter via the onEdit callback, and this must be set to true to enable.
   */
  editMode?: boolean;
  /**
   * Callback fired whenever the component wishes to enter edit mode.
   */
  onEdit?: () => void;
  /**
   * Callback fired whenever the component wishes to update the value.
   */
  onSave?: (newValue: string) => void;
  /**
   * Callback fired whenever the component wishes to cancel edit mode without
   * saving the new value.
   */
  onCancel?: (originalValue: string) => void;
  /**
   * Classname for outermost div.
   */
  className?: string;
  /**
   * Is the editable text disabled (will not be editable).
   */
  disabled?: boolean;
  /**
   * Render out a custom string to display when not in edit mode.
   */
  onRenderText?: (value: string) => string;
  /**
   * A placeholder to display if there's no value. This will also account for
   * `onRenderText`, so if you want to display this then `onRenderText` must
   * return a falsy value like empty string.
   */
  placeholder?: string;
  /**
   * Center the text within the editable text field.
   */
  centerText?: boolean;
}

export const EditableText = forwardRef<HTMLDivElement, EditableTextProps>(
  (
    {
      value: controlledValue,
      editMode: controlledEditMode,
      onEdit,
      onSave,
      onCancel,
      className,
      disabled,
      onRenderText,
      placeholder,
      centerText,
    },
    ref,
  ) => {
    // Use state if controlled edit mode not provided.
    const [stateEditMode, setStateEditMode] = useState(false);
    const editMode = controlledEditMode ?? stateEditMode;

    // Use state if controlled value not provided.
    const [stateValue, setStateValue] = useState('');
    const value = controlledValue ?? stateValue;

    // Keep edit value in sync with value (controlled or state) and reset when
    // edit mode is cancelled.
    const [editValue, setEditValue] = useState(value);
    useEffect(() => setEditValue(value), [editMode, value]);

    // Focus input whenever edit mode is enabled.
    const inputRef = useRef<HTMLInputElement>(null);
    useEffect(() => {
      if (editMode) inputRef.current?.focus();
    }, [editMode]);

    const handleOnEdit = useCallback(() => {
      if (controlledEditMode === undefined) setStateEditMode(true);
      onEdit?.();
    }, [onEdit, controlledEditMode]);

    const handleOnCancel = useCallback(() => {
      if (controlledEditMode === undefined) setStateEditMode(false);
      setEditValue(value);
      onCancel?.(value);
    }, [controlledEditMode, onCancel, value]);

    const handleOnSave = useCallback(() => {
      if (controlledValue === undefined) setStateValue(editValue);
      if (controlledEditMode === undefined) setStateEditMode(false);
      onSave?.(editValue);
    }, [controlledEditMode, controlledValue, editValue, onSave]);

    const handleKeyUp = useCallback(
      (event: KeyboardEvent<HTMLInputElement>) => {
        if (event.key === 'Escape') handleOnCancel();
        if (event.key === 'Enter') handleOnSave();
      },
      [handleOnSave, handleOnCancel],
    );

    const [valueToDisplay, isPlaceholder] = useMemo(() => {
      const renderedValue = onRenderText ? onRenderText(value) : value;
      if (placeholder && !renderedValue) return [placeholder, true];
      return [renderedValue, false];
    }, [onRenderText, placeholder, value]);

    const editableTextClass = classnames(
      'ui__base ui__editableText',
      { ['ui__editableText--disabled']: disabled },
      { ['ui__editableText--centerText']: centerText },
      className,
    );

    const buttonClass = classnames(
      'ui__editableText__button',
      { ['ui__editableText__button--placeholder']: isPlaceholder },
      className,
    );

    return (
      <div className={editableTextClass}>
        {editMode ? (
          <input
            className="ui__editableText__field"
            value={editValue}
            onChange={e => setEditValue(e.target.value)}
            onKeyUp={handleKeyUp}
            ref={inputRef}
            onBlur={handleOnSave}
          />
        ) : (
          <ClickableDiv className={buttonClass} disabled={disabled} onClick={handleOnEdit} ref={ref} usePointer>
            {valueToDisplay}
          </ClickableDiv>
        )}
      </div>
    );
  },
);
