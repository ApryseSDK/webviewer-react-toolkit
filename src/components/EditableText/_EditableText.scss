.ui__editableText {
  &--disabled {
    .ui__editableText__button {
      color: $blue-gray-2;
    }
  }

  &--centerText {
    .ui__editableText__button {
      justify-content: center;
    }

    .ui__editableText__field {
      text-align: center;
    }
  }

  &__button,
  &__field {
    min-height: 28px;
    min-width: 100%;
    font-weight: 400;
    color: $font-primary;
    padding: $padding-tiny;
    border-radius: $border-radius;
    border: 1px solid $blue-gray-1;
  }

  &__button {
    display: flex;
    align-items: center;
  }

  &:not(.ui__editableText--disabled) .ui__editableText__button {
    &:hover {
      background: $blue-gray-1;
    }

    &--placeholder {
      color: $blue-gray-3;
    }
  }

  &__field {
    background: $blue-gray-1;
    outline: none;

    &:focus {
      @include focus;
    }
  }
}
