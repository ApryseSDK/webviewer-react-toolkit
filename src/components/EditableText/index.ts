export { EditableText as default, EditableTextProps } from './EditableText';
