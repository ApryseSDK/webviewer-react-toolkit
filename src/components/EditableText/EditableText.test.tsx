import { shallow } from 'enzyme';
import React from 'react';
import { EditableText } from '../EditableText';

describe('EditableText component', () => {
  it('renders its contents', () => {
    const editableText = shallow(<EditableText />);
    expect(editableText.find('.ui__editableText')).toHaveLength(1);
  });

  it('snapshot renders default editableText', () => {
    const editableText = shallow(<EditableText />);
    expect(editableText).toMatchSnapshot();
  });
});
