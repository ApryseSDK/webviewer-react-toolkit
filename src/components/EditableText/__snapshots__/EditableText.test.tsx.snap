// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableText component snapshot renders default editableText 1`] = `
<div
  className="ui__base ui__editableText"
>
  <ForwardRef
    className="ui__editableText__button"
    onClick={[Function]}
    usePointer={true}
  />
</div>
`;
